### PR TITLE
Codegangsta to urfave cli

### PR DESCRIPTION
**Pull request by GrantSeltzer**
- Updated the name of the cli library from codegangsta to urfave (Moved)
- Upgraded the library
- Changed definitions of functions that are used as app actions to reflect the library changes. Now these functions must return errors. We're still using logrus but return nils at end of the functions.

### CODE REVIEW

**Comment by runcom** [comment]
Ca you put your second commit here on top of the commit when have in #113? This pr is missing the actual vendoring process (./hack/vendor.sh call) while #113 is ok, then you can squash the two commits so git bisect is fine

File: cmd/skopeo/layers.go [hunk 1]
```diff
@@ -2,20 +2,23 @@ package main
...
 // TODO(runcom): document args and usage
 var layersCmd = cli.Command{
 	Name:  "layers",
 	Usage: "get images layers",
-	Action: func(c *cli.Context) {
+	Action: func(c *cli.Context) error {
 		img, err := parseImage(c)
 		if err != nil {
 			logrus.Fatal(err)
+			return err
```
Comment (runcom) [inline cmd/skopeo/layers.go]: is this required? isn't Fatal (like Fatalf) just stop execution? (also below)

File: cmd/skopeo/layers.go [hunk 2]
```diff
@@ -2,20 +2,23 @@ package main
...
 		}
 		if err := img.Layers(c.Args().Tail()...); err != nil {
 			logrus.Fatal(err)
+			return err
```
Comment (runcom) [inline cmd/skopeo/layers.go]: ditto

File: cmd/skopeo/inspect.go [hunk 3]
```diff
@@ -45,7 +45,6 @@ var inspectCmd = cli.Command{
 		}
 		if c.Bool("raw") {
 			fmt.Println(string(rawManifest))
-			return
```
Comment (runcom) [inline cmd/skopeo/inspect.go]: mmm this is wrong, here you need to `return nil`

**Comment by GrantSeltzer** [comment]
LGTM!

**Comment by runcom** [comment]
@mtrmac want to have a look at this?

**Comment by mtrmac** [comment]
Sure, tests pass, and hopefully somebody has checked that the commands still seem to actually work…

And I very much like that this version of `cli` actually supports unit testing without adding ugly wrappers in #16:

 - There is now an `App.Writer` and `App.ErrWriter`
 - `Action` handlers can return an error (turned into `logrus.Fatal` at the top-level `main`) instead of calling `logrus.Fatalf` immediately.

Grant, if you still have the patch which turns `logrus.Fatal` into `return` in the various places, I would love to have that.  (If not, no harm done, I can redo that when rebasing #16.)

[Oh, and now working on the integration tests, I am much more confident in thinking that #16 is desirable. The edit-compile-test cycle time for integration tests is pretty horrible, so the basic error handling and trivial local cases like `dir:` should run as local unit tests.]

**Comment by runcom** [comment]
@mtrmac great!!

**Comment by runcom** [comment]
@GrantSeltzer can you do the `logrus.Fatal[f]` to `return err` conversion as you did before (but we dropped that just not to change that much)?

Feel free to add another commit on top of this (instead of amending to the first one)

**Comment by GrantSeltzer** [comment]
@runcom I'm on it

**Comment by GrantSeltzer** [comment]
@runcom @mtrmac What are your thoughts on instances that we use logrus.Fatalf? for example:

`			logrus.Fatalf("Error computing manifest digest: %s", err.Error())
`
Keep those? Just return the errror? or Concatenate the errors and return that?

**Comment by mtrmac** [comment]
@GrantSeltzer 

Yes, or
```golang
return fmt.Errorf("Error computing manifest digest: %s", err.Error())
```
which would be a smaller change.  Either works fine, yours is a bit more efficient.

Just take care to preserve the `": "` to keep the output readable, please.

**Comment by runcom** [comment]
> return fmt.Errorf("Error computing manifest digest: %s", err.Error())

right just remove `.Error()` here and use either %s or %v for errors (I like %v for errors, we do that in docker as well)

**Comment by runcom** [comment]
Sorry guys, can we do this with `fmt.Errorf` and using either `%s` or `%v` w/o `.Error()` calls?

**Comment by runcom** [comment]
FWIW `return errors.New(("Error initializing " + string(context.Args()[0])) + err.Error())` I don't like this
just
```
return fmt.Errorf("Error initializing %s: %v", context.Args()[0], err)
```
is far better to me..

**Comment by GrantSeltzer** [comment]
Sure, will change as soon as I get home

File: cmd/skopeo/copy.go [hunk 4]
```diff
@@ -41,70 +41,70 @@ type manifestSchema1 struct {
 
 func copyHandler(context *cli.Context) error {
 	if len(context.Args()) != 2 {
-		logrus.Fatal("Usage: copy source destination")
+		return fmt.Errorf("Usage: copy source destination")
```
Comment (runcom) [inline cmd/skopeo/copy.go]: w/o formatting, this would be just `errors.New("")`

Comment (GrantSeltzer) [inline cmd/skopeo/copy.go]: So would this be better as logrus.Fatal?

Comment (runcom) [inline cmd/skopeo/copy.go]: nope,it would be:
```
return errors.New("Usage: copy source destination")
```

File: image/image.go [hunk 5]
```diff
@@ -238,7 +237,7 @@ func fixManifestLayers(manifest *manifestSchema1) error {
 		}
 	}
 	if imgs[len(imgs)-1].Parent != "" {
-		return errors.New("Invalid parent ID in the base layer of the image.")
+		return fmt.Errorf("Invalid parent ID in the base layer of the image.")
```
Comment (runcom) [inline image/image.go]: nitpicking - this was good with errors.New :p

Comment (GrantSeltzer) [inline image/image.go]: @runcom :+1:

**Pull request merged**